src/caffe/CMakeLists.txt

```diff
@@ -21,12 +22,40 @@ add_library(caffe ${srcs})
 target_link_libraries(caffe proto ${Caffe_LINKER_LIBS})
 caffe_default_properties(caffe)
 
+# ---[ Visual Studio specifics
+#  configure headers for dynamic or static linking on windows
+unset(CAFFE_DLLEXPORT)
+unset(CAFFE_DLLIMPORT)
+unset(CAFFE_INCLUDE_FORCELINK)
+if(MSVC)
+    if(BUILD_SHARED_LIBS)
+        set(CAFFE_DLLEXPORT __declspec(dllexport))
+        set(CAFFE_DLLIMPORT __declspec(dllimport))
+    else()
+        get_filename_component(forcelink_header_name ${forcelink_header} NAME)
+        set(CAFFE_INCLUDE_FORCELINK "#      include \"caffe/${forcelink_header_name}\"")
+        # add a post build command to generate a header file that will
+        # force the linker to include the statically registered layers
+        # in a final executable.
+        include(${PROJECT_SOURCE_DIR}/cmake/GenerateForceLinkSymbolsHeader.cmake)    
+        add_generate_force_link_symbols_header_post_build_command(caffe ${forcelink_header} caffe_force_link)        
+        # install the generated file
+        install(FILES ${forcelink_header} DESTINATION include/caffe)
+    endif()        
+endif()
+
+configure_file("${PROJECT_SOURCE_DIR}/cmake/Templates/export.hpp.in" 
+                ${export_header} @ONLY)
+# define preprocessor macro so that we will not include the generated forcelink header
+target_compile_definitions(caffe PRIVATE -DBUILDING_CAFFE_LIB)
```

Comment:
Warning. This cmake command is available since v. 2.8.11.  For now minimal required cmake is 2.8.7, which is apt-get-installed by default on Ubuntu 12.04. Seems the change to be discussed with caffe authors/community.